extensions/wikia/CategoryPagination/CategoryPaginationViewer.class.php

```diff
@@ -2,102 +2,226 @@
+		 */
+		$categoryGallery = '';
+		wfRunHooks( 'CategoryPage::getCategoryTop', [ $this, &$categoryGallery ] );
+
+		// Render
+		$this->getOutput()->addHeadItem( 'Paginator', $this->getPaginationHeadItem() );
+		$tpl = new EasyTemplate( __DIR__ . '/templates' );
+		$tpl->set_vars( [
+			'aswersSection' => $answersSection,
+			'categoryGallery' => $categoryGallery,
+			'content' => $this->content,
+			'counts' => $this->counts,
+			'empty' => empty( $answersSection ) && empty( $this->content ),
+			'lang' => $this->getLanguage(),
+			'numberShown' => $this->numberShown,
+			'paginator' => $this->paginators,
+			'titleEscaped' => htmlspecialchars( $this->getTitle()->getText() ),
+		] );
+		return $tpl->render( 'category' );
+	}
 
-		$cat = Category::newFromTitle( $this->title );
+	/**
+	 * Generate the data for each of the default sections
+	 * Save it to $this->content and $this->numberShown
+	 */
+	private function generateData() {
+		// Get counts
+		$cat = $this->getCat();
 		$this->counts = [
-			'file' => $cat->getFileCount(),
-			'subcat' => $cat->getSubcatCount(),
-			'page' => $cat->getPageCount() - $cat->getFileCount() - $cat->getSubcatCount(),
+			self::TYPE_FILE => $cat->getFileCount(),
+			self::TYPE_SUBCAT => $cat->getSubcatCount(),
+			self::TYPE_PAGE => $cat->getPageCount() - $cat->getFileCount() - $cat->getSubcatCount(),
 		];
 
-		$url = $this->title->getLocalURL();
-
-		foreach ( array_keys( $this->from ) as $type ) {
-			$this->from[$type] = $this->getActualFrom( $type, $page );
-			$this->paginators[$type] = new Paginator( $this->counts[$type], $this->limit, $url );
+		// Set up pagination
+		$url = $cat->getTitle()->getFullURL();
+		$this->page = max( 1, $this->getRequest()->getInt( 'page', 1 ) );
+		foreach ( $this->counts as $type => $count ) {
+			$this->paginators[$type] = new Paginator( $count, $this->limit, $url );
 			$this->paginators[$type]->setActivePage( $this->page );
 		}
 
-		$context->getOutput()->addHeadItem( 'Paginator', $this->getPaginationHeadItem() );
-	}
-
-	private function getPaginationHeadItem() {
-		$numPages = -1;
-		$maxPaginator = null;
-		foreach ( $this->paginators as $paginator ) {
-			if ( $paginator->getPagesCount() > $numPages ) {
-				$numPages = $paginator->getPagesCount();
-				$maxPaginator = $paginator;
+		// Get pages
+		$this->processSection( self::TYPE_PAGE, function ( $title, $humanSortkey, $row ) {
+			/** @see answerAddCategoryPage */
+			if ( wfRunHooks( 'CategoryViewer::addPage', array( &$this, &$title, &$row, $humanSortkey ) ) ) {
+				$this->addPage( $title, $humanSortkey, $row->page_len, $row->page_is_redirect );
+				return true;
 			}
-		}
-		if ( $maxPaginator ) {
-			return $maxPaginator->getHeadItem();
-		}
-		return '';
+			return false;
+		}, function () {
+			return $this->formatList( $this->articles, $this->articles_start_char );
+		} );
+
+		// Get media
+		$this->processSection( self::TYPE_FILE, function ( $title, $humanSortkey, $row ) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Use "static" keyword instead of "self". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2037)